### PR TITLE
Implementation of release-automation exception process

### PR DESCRIPTION
**Pull request by tomaszbarwicki**
PR to add exceptions mechanism for release-automation tool. The implementations allows processing TRG exceptions within automated checks based on submitted requests. 

The work item contains also exceptions process description, data for already known exceptions, checks reordering to match TRG sequence and formatting.

Updates https://github.com/eclipse-tractusx/sig-infra/issues/425

### CODE REVIEW

**Comment by tomaszbarwicki** [comment]
UT TestShouldPassIfExceptionDataIsAtURL fails because the file doesn't exists yet, it's part of the PR items.

**Comment by almadigabor** [comment]
Hey @tomaszbarwicki! Great work here. One remark that the code block below is repeated many times in different quality checks. Is there a way to refactor it to be a reusable function in all files?

```go
	config, err := exception.GetData()
	if err != nil {
		log.Println("Can't process exceptions.")
	} else {
		repoInfo := repo.GetRepoBaseInfo(a.baseDir)
		if config.IsExceptioned(a.Name(), "https://github.com/eclipse-tractusx/"+repoInfo.Reponame) {
			return &tractusx.QualityResult{Passed: true}
		}
	}
```

**Comment by tomaszbarwicki** [comment]
> Hey @tomaszbarwicki! Great work here. One remark that the code block below is repeated many times in different quality checks. Is there a way to refactor it to be a reusable function in all files?
> 
> ```go
> config, err := exception.GetData()
> if err != nil {
> 	log.Println("Can't process exceptions.")
> } else {
> 	repoInfo := repo.GetRepoBaseInfo(a.baseDir)
> 	if config.IsExceptioned(a.Name(), "https://github.com/eclipse-tractusx/"+repoInfo.Reponame) {
> 		return &tractusx.QualityResult{Passed: true}
> 	}
> }
> ```

Thanks for feedback @almadigabor, I made the changes and moved the code outside the checks into command line runner and dashboard, hope that looks better now :). Please review.

**Review by almadigabor — DISMISSED** [review]

**Review by SebastianBezold — CHANGES_REQUESTED** [review]
Some things I would like to see changed. 
Additionally, I think doing a string compare on the TRG name to find out, if there is an exception for it, could become quite error prone. When refactoring, or renaming, we need to make sure, it is done in both places. An alternative option could be to introduce IDs, since they would potentially be treated as "do not change". 

But fine for now with comparing the name

File: release-automation/EXCEPTIONS.md [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+## Objective
+
+The process aims to establish a structured method for raising exceptions to the Tractus-X Release Guidelines (TRG) for the release automation tool. These exceptions are managed through pull requests (PRs) modifying the `exceptions.yaml` file. The system team will review these exceptions and make decisions accordingly.
+
+1. **Identification of Exception:**
+   - Before raising an exception, the requester must identify a specific guideline that needs to be exempted.
+
+2. **Understanding Structure:**
+   - Review the structure of the [exceptions.yaml](https://github.com/eclipse-tractusx/sig-release/blob/main/release-automation/internal/exception/exceptions.yaml) file to understand how exceptions are documented. Ensure familiarity with the format, including key-value pairs and any required fields. Proper exception needs to have repository url under respective TRG.
```
[the remark this answers is not in the document]
Comment (tomaszbarwicki) [reply release-automation/EXCEPTIONS.md]: corrected: https://github.com/eclipse-tractusx/sig-release/pull/558/commits/74bfe3050a73eb45bd095a36b2fdccb698dad138

File: release-automation/EXCEPTIONS.md [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+4. **Modifying `exceptions.yaml`:**
+   - Navigate to the [exceptions.yaml](https://github.com/eclipse-tractusx/sig-release/blob/main/release-automation/internal/exception/exceptions.yaml) file within the repository.
+   - Modify the file to include the necessary exception(s) following the predefined structure.
+
+5. **Drafting a Pull Request (PR):**
+   - Create a PR to propose the changes made to the [exceptions.yaml](https://github.com/eclipse-tractusx/sig-release/blob/main/release-automation/internal/exception/exceptions.yaml) file.
+   - Provide a clear and descriptive title for the PR, summarizing the purpose of the exception(s).
+   - In the PR description, elaborate on the reasons for requesting the exception(s) and any potential impact on the release process.
+
+6. **Assigning Reviewers:**
```
[the remark this answers is not in the document]
Comment (tomaszbarwicki) [reply release-automation/EXCEPTIONS.md]: shortened: https://github.com/eclipse-tractusx/sig-release/pull/558/commits/74bfe3050a73eb45bd095a36b2fdccb698dad138

File: release-automation/internal/dashboard/tractusx.go [hunk 3]
```diff
@@ -109,16 +110,31 @@ func repoClone(repo Repository) string {
...
 	}
 
 	for _, check := range initializeChecksForDirectory(dir) {
-		testResult := check.Test()
+		config, err := exception.GetData()
+		exceptionPresent := false
+		if err != nil {
+			log.Println("Can't process exceptions.")
+		} else {
+			repoInfo := repo.GetRepoBaseInfo(check.BaseDir())
```
[the remark this answers is not in the document]
Comment (tomaszbarwicki) [reply release-automation/internal/dashboard/tractusx.go:128]: @SebastianBezold: Indeed that would work for dashboard build run, however for CLI run I don't think I have other option to get the repo directory hence the getter added to interface.

File: release-automation/internal/exception/exception.go [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+type Exception struct {
+	Trg          string   `yaml:"trg"`
+	Repositories []string `yaml:"repository"`
+}
+
+type Config struct {
+	Exceptions []Exception `yaml:"exceptions"`
+}
+
+func GetData() (*Config, error) {
```
[the remark this answers is not in the document]
Comment (tomaszbarwicki) [reply release-automation/internal/exception/exception.go]: Made it ExceptionsMap return type:
https://github.com/eclipse-tractusx/sig-release/blob/5336a436e14d726b360280c7c0f030536d1b139f/release-automation/internal/exception/exception.go#L42

File: release-automation/internal/exception/exception.go [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+	var c Config
+	err := yaml.Unmarshal(data, &c)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse YAML data")
+	}
+	return &c, nil
+}
+
+// Checks if repository has exception for specific TRG
+func (c *Config) IsExceptioned(trg string, repository string) bool {
```
[the remark this answers is not in the document]
Comment (tomaszbarwicki) [reply release-automation/internal/exception/exception.go]: Reduced the loops by the map usage:
https://github.com/eclipse-tractusx/sig-release/blob/5336a436e14d726b360280c7c0f030536d1b139f/release-automation/internal/exception/exception.go#L81-L86

File: release-automation/internal/dashboard/tractusx.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (SebastianBezold) [reply release-automation/internal/dashboard/tractusx.go]: Hmm ok, I guess we can live with that, even though I think we could find a better solutions in the future

**Review by SebastianBezold — APPROVED** [review]
👍

**Pull request merged**